go/consensus/cometbft/beacon/beacon.go

```diff
@@ -101,47 +106,66 @@ func (sc *serviceClient) GetFutureEpoch(ctx context.Context, height int64) (*bea
 
 func (sc *serviceClient) GetEpochBlock(ctx context.Context, epoch beaconAPI.EpochTime) (int64, error) {
 	now, currentBlk := sc.currentEpochBlock()
-	if epoch == now {
-		return currentBlk, nil
-	}
-
-	// The epoch can't be earlier than the initial starting epoch.
 	switch {
+	case epoch == now:
+		_ = sc.epochCache.Put(epoch, currentBlk)
+		return currentBlk, nil
 	case epoch < sc.baseEpoch:
 		return -1, fmt.Errorf("epoch predates base (base: %d requested: %d)", sc.baseEpoch, epoch)
 	case epoch == sc.baseEpoch:
 		return sc.baseBlock, nil
 	}
 
+	// Try the cache first.
+	if cachedHeight, ok := sc.epochCache.Get(epoch); ok {
+		return cachedHeight.(int64), nil
+	}
+
+	lowHeight, err := sc.backend.GetLastRetainedVersion(ctx)
+	if err != nil {
+		return -1, fmt.Errorf("failed to query last retained version: %w", err)
+	}
+
+	blk, err := sc.backend.GetCometBFTBlock(ctx, consensus.HeightLatest)
+	if err != nil {
+		return -1, err
+	}
+	hiHeight := blk.Height
+	// Start with the latest height as it is possible that currentEpochBlock is not the most up to
+	// date in cases where GetEpochBlock is called during epoch transitions.
+	height := hiHeight
+
 	// Find historic epoch.
-	//
-	// TODO: This is really really inefficient, and should be optimized,
-	// maybe a cache of the last few epochs, or a binary search.
-	height := consensus.HeightLatest
-	for {
+	const maxIterations = 20 // Should be good enough for most use cases.
+	for range maxIterations {
```

Comment:
Sure, I thought of that as well, but even doing a few steps is much better than the previous O(n) implementation. If we see this is still a problem we can optimize further.

Reply:
Since bisection is done on `int64` (max 64 steps), this limit could also be removed. But then the stop condition should be improved (one should check if the low/high epoch hasn't changed). Currently, if we use a very high value for epoch (value greater than the last epoch), we will do 20 steps always, and always querying the same height .

Reply:
> I thought of that as well, but even doing a few steps is much better than the previous O(n) implementation

Agree, unless these few steps need to query the database or fetch blocks from other nodes. In that case, every removed step is a major improvement. Not sure if this is the case here.

Reply:
This is querying the local database, so it should be pretty fast for a few queries. Previously the problem was if one was querying an epoch very far in the past (there is a use case for this in the CLI) this did a lot of queries.

Reply:
Added an early stopping condition.